Exchange Head single-fligth protection

IliaBulavintsev opened the pull request:
https://github.com/celestiaorg/go-header/issues/154
## Overview

This PR adds singleflight protection to prevent duplicate header requests in the Exchange component. The focus is on minimizing redundant requests, especially around the Head method, which is most likely to experience simultaneous calls.

renaynay reviewed (commented):
This is a great start :) just a few things 

* can you please remove the `sync_getter` from the syncer component as now the single flight protection has moved to the exchange instead? 
* re tests: they are comprehensive and have good coverage, but I think they may be a bit overkill. Testing 2 to max 3 simultaneous requests would be sufficient. Also the `requestFromTrusted` param in the tests would be more useful in terms of coverage if it was false for some cases.
* nit: the tests could benefit from a bit of spacing between statements for readability :)

IliaBulavintsev commented:
Hey Rene!

- The `sync_getter` was removed as the single flight protection has been moved to the Exchange.
- The test cases were also adjusted—you're right about the unnecessary number of cases, were reduced. Also added a test case with an untrusted peer and removed the untrusted peer instance that was running.

Lmk if you have other feedback

renaynay commented on p2p/exchange_test.go:
```diff
@@ -156,6 +157,70 @@ func TestExchange_RequestHead_UnresponsivePeer(t *testing.T) {
+		expectedHeight     uint64
+		expectedHash       header.Hash
+	}{
+		{
+			requestFromTrusted: true,
+			lastHeader:         trustedStore.Headers[trustedStore.HeadHeight-1],
+			expectedHeight:     trustedStore.HeadHeight,
+			expectedHash:       trustedStore.Headers[trustedStore.HeadHeight].Hash(),
+		},
+		{
+			requestFromTrusted: true,
+			lastHeader:         trustedStore.Headers[trustedStore.HeadHeight-1],
+			expectedHeight:     trustedStore.HeadHeight,
+			expectedHash:       trustedStore.Headers[trustedStore.HeadHeight].Hash(),
+		},
```
what's the difference between these two test cases?

IliaBulavintsev replied:
The duplicate cases serve to double-check that `singleflight` consistently returns the same response for identical requests from a `trusted` source. However, if you think a single case is sufficient for validation, we can remove the duplicates to streamline the test

renaynay replied:
That seems like a good test generally-speaking but I don't think it is relevant to the single-flight protection feature and should rather be implemented in a different way. I appreciate the thought here, however.

IliaBulavintsev replied:
Got it, I’ll remove the test as it’s not directly relevant to single-flight. Thanks for the feedback!

Wondertan commented:
Taking over this PR.

I moved out single-flight protection out of `exchange` to be in `sync.Syncer`. This way the single-flight protection works for every `Head` implementation, like we have in `core`. Besides, I avoid using the `x/sync/singleflight` as it does not handle context cancellation.

cristaloleg commented on sync/head_sync.go:
```diff
@@ -0,0 +1,51 @@
+
+import (
+	"context"
+	"github.com/celestiaorg/go-header"
```
unsorted, please run `make fmt`.

Wondertan replied:
once we base it on #256

Wondertan replied:
done

Wondertan commented on sync/head_sync_test.go:
```diff
@@ -0,0 +1,44 @@
+
+	wg.Wait()
+
+	assert.Less(t, int(counter.cntr.Load()), callsN)
```
What do you suggest?

cristaloleg replied:
Like 70% of calls are cached or so.

This might involve changes in delay formula.

Wondertan commented on sync/head_sync.go:
```diff
@@ -0,0 +1,51 @@
+)
+
+// syncHead is a wrapper around the header.Head that provides single-flight access to the Head.
+// TODO: Should we extract this as public component?
```
The question is: who should be responsible for single-flight protection, Syncer or the Exchange? Currently it is Syncer and its stays so with this PR. Extracting this to a public component may help Exchange implementers if this make them responsible for single-flight protection.

Wondertan replied:
Leaning towards keeping exchanges lean so they don't need to worry about single flight.

cristaloleg commented:
Hey @IliaBulavintsev are you going to address recent comments?
